File: src/workspace.rs
// Workspace file CRUD, ported from the Tauri backend (src-tauri/src/lib.rs) in the
// original OfficeSuite desktop app. There, `root` came straight from the client
// because it was gated by an OS folder-picker dialog for a single local user. On
// the web there is no such gate, so `root` is instead always derived from the
// authenticated session (see `current_user_id`), and every relative path is
// validated with `safe_rel_path` to make sure it can't escape that root.

use axum::extract::{Query, State};
use axum::{http::HeaderMap, http::StatusCode, Json};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::fs;
use std::path::{Component, Path, PathBuf};
use std::time::UNIX_EPOCH;
use tower_sessions::Session;

use crate::auth::current_user_id_with_headers;
use crate::error::AppError;
use crate::state::AppState;

#[derive(Serialize)]
pub struct FsEntry {
    name: String,
    #[serde(rename = "relPath")]
    rel_path: String,
    #[serde(rename = "isDir")]
    is_dir: bool,
    modified: u64,
    /// sha256 hex digest of the file's bytes, empty string for directories.
    /// Lets sync clients tell "did this change on the server" from one
    /// listing call without downloading every file's content.
    hash: String,
    children: Vec<FsEntry>,
}

fn hash_file(path: &Path) -> String {
    match fs::read(path) {
        Ok(bytes) => {
            let mut hasher = Sha256::new();
            hasher.update(&bytes);
            format!("{:x}", hasher.finalize())
        }
        Err(_) => String::new(),
    }
}

pub const WORK_FILE_EXTENSIONS: [&str; 8] = ["mdp", "mds", "mdg", "mdn", "mdl", "mdc", "mde", "mdb"];

/// Root-level sidecar files that sync clients need to see in listings (with a
/// hash) even though they aren't work files: the user's custom document
/// templates and roaming UI preferences (theme), so both follow the account
/// across devices. The web UI filters these out of its file tree; the desktop
/// tree never lists them (its walker only shows work extensions).
const SYNC_SIDECAR_FILES: [&str; 2] = ["_lktpl.json", "_lkprefs.json"];

fn modified_ms(metadata: &fs::Metadata) -> u64 {
    metadata
        .modified()
        .ok()
        .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

/// Rejects absolute paths, drive-letter prefixes, and `..` components so a
/// request can never resolve to a path outside the caller's workspace root.
pub fn safe_rel_path(rel_path: &str) -> Result<PathBuf, AppError> {
    let mut buf = PathBuf::new();
    for component in Path::new(rel_path).components() {
        match component {
            Component::Normal(seg) => buf.push(seg),
            Component::CurDir => {}
            _ => return Err(AppError::BadRequest("invalid path".to_string())),
        }
    }
    if buf.as_os_str().is_empty() {
        return Err(AppError::BadRequest("invalid path".to_string()));
    }
    Ok(buf)
}

/// Total bytes stored under a workspace root. Walks the tree rather than
/// tracking a running total in the DB: workspaces are small (text files) and
/// this way the number can't drift out of sync with what's on disk.
fn dir_size_bytes(dir: &Path) -> u64 {
    let Ok(read) = fs::read_dir(dir) else { return 0 };
    read.flatten()
        .map(|entry| match entry.metadata() {
            Ok(m) if m.is_dir() => dir_size_bytes(&entry.path()),
            Ok(m) => m.len(),
            Err(_) => 0,
        })
        .sum()
}

/// Registration is open, so an unbounded workspace lets any account fill the
/// server's disk. Checked before a write that would grow the workspace;
/// replacing a file with a smaller one always succeeds, so a user who hits
/// the ceiling can still edit their way back under it.
pub fn check_quota(
    state: &AppState,
    root: &Path,
    target: &Path,
    incoming_len: u64,
) -> Result<(), AppError> {
    if state.workspace_quota_bytes == 0 {
        return Ok(());
    }
    let existing_len = fs::metadata(target).map(|m| m.len()).unwrap_or(0);
    if incoming_len <= existing_len {
        return Ok(());
    }
    let projected = dir_size_bytes(root) - existing_len + incoming_len;
    if projected > state.workspace_quota_bytes {
        return Err(AppError::BadRequest(format!(
            "workspace is full ({} MB max) — delete something first",
            state.workspace_quota_bytes / (1024 * 1024)
        )));
    }
    Ok(())
}

async fn user_ctx(
    state: &AppState,
    session: &Session,
    headers: &HeaderMap,
) -> Result<(String, PathBuf), AppError> {
    let user_id = current_user_id_with_headers(state, session, headers).await?;
    let root = state.workspaces_dir.join(&user_id);
    fs::create_dir_all(&root)?;
    Ok((user_id, root))
}

async fn user_root(
    state: &AppState,
    session: &Session,
    headers: &HeaderMap,
) -> Result<PathBuf, AppError> {
    Ok(user_ctx(state, session, headers).await?.1)
}

// ── Deletion tombstones (see migrations/0007_deleted_files.sql) ──

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// Canonical tombstone key for a validated rel path: forward-slash separators,
/// matching the rel-path form clients use in every request and listing.
fn rel_key(rel: &Path) -> String {
    rel.components()
        .map(|c| c.as_os_str().to_string_lossy())
        .collect::<Vec<_>>()
        .join("/")
}

async fn record_tombstone(state: &AppState, user_id: &str, rel_path: &str) {
    let _ = sqlx::query(
        "INSERT INTO deleted_files (user_id, rel_path, deleted_at) VALUES (?, ?, ?)
         ON CONFLICT(user_id, rel_path) DO UPDATE SET deleted_at = excluded.deleted_at",
    )
    .bind(user_id)
    .bind(rel_path)
    .bind(now_ms())
    .execute(&state.db)
    .await;
}

async fn clear_tombstone(state: &AppState, user_id: &str, rel_path: &str) {
    let _ = sqlx::query("DELETE FROM deleted_files WHERE user_id = ? AND rel_path = ?")
        .bind(user_id)
        .bind(rel_path)
        .execute(&state.db)
        .await;
}

/// Every file (not directory) under `dir`, as rel keys prefixed with
/// `rel_prefix` — used to tombstone the contents of a deleted/moved folder.
fn collect_files_under(dir: &Path, rel_prefix: &str, out: &mut Vec<String>) {
    let Ok(read) = fs::read_dir(dir) else { return };
    for entry in read.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        let rel = if rel_prefix.is_empty() {
            name
        } else {
            format!("{}/{}", rel_prefix, name)
        };
        if entry.path().is_dir() {
            collect_files_under(&entry.path(), &rel, out);
        } else {
            out.push(rel);
        }
    }
}

pub fn walk_work_dir(dir: &Path, rel_prefix: &str) -> Result<Vec<FsEntry>, AppError> {
    let mut entries = Vec::new();

    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        let name = entry.file_name().to_string_lossy().to_string();
        let metadata = entry.metadata()?;

        let rel_path = if rel_prefix.is_empty() {
            name.clone()
        } else {
            format!("{}/{}", rel_prefix, name)
        };

        if metadata.is_dir() {
            entries.push(FsEntry {
                name,
                children: walk_work_dir(&path, &rel_path)?,
                rel_path,
                is_dir: true,
                modified: 0,
                hash: String::new(),
            });
        } else {
            let ext = Path::new(&name)
                .extension()
                .and_then(|e| e.to_str())
                .map(|e| e.to_lowercase())
                .unwrap_or_default();

            let is_root_sidecar =
                rel_prefix.is_empty() && SYNC_SIDECAR_FILES.contains(&name.as_str());
            if !WORK_FILE_EXTENSIONS.contains(&ext.as_str()) && !is_root_sidecar {
                continue;
            }

            entries.push(FsEntry {
                name,
                rel_path,
                is_dir: false,
                modified: modified_ms(&metadata),
                hash: hash_file(&path),
                children: Vec::new(),
            });
        }
    }

    entries.sort_by(|a, b| match (a.is_dir, b.is_dir) {
        (true, false) => std::cmp::Ordering::Less,
        (false, true) => std::cmp::Ordering::Greater,
        _ => a.name.to_lowercase().cmp(&b.name.to_lowercase()),
    });

    Ok(entries)
}

pub async fn list_work_folder(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
) -> Result<Json<Vec<FsEntry>>, AppError> {
    let root = user_root(&state, &session, &headers).await?;
    Ok(Json(walk_work_dir(&root, "")?))
}

#[derive(Deserialize)]
pub struct PathQuery {
    path: String,
}

/// Cheap change-detection for one file: lets a client poll "did anyone else
/// write this?" without downloading the content or walking the whole tree.
#[derive(Serialize)]
pub struct FileStat {
    modified: u64,
    hash: String,
}

pub fn stat_file(path: &Path) -> Result<FileStat, AppError> {
    let metadata = fs::metadata(path)?;
    if !metadata.is_file() {
        return Err(AppError::NotFound);
    }
    Ok(FileStat {
        modified: modified_ms(&metadata),
        hash: hash_file(path),
    })
}

pub async fn stat_work_file(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Query(q): Query<PathQuery>,
) -> Result<Json<FileStat>, AppError> {
    let root = user_root(&state, &session, &headers).await?;
    let rel = safe_rel_path(&q.path)?;
    Ok(Json(stat_file(&root.join(rel))?))
}

pub async fn read_work_file(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Query(q): Query<PathQuery>,
) -> Result<String, AppError> {
    let root = user_root(&state, &session, &headers).await?;
    let rel = safe_rel_path(&q.path)?;
    fs::read_to_string(root.join(rel)).map_err(|e| e.into())
}

pub async fn write_work_file(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Query(q): Query<PathQuery>,
    content: String,
) -> Result<StatusCode, AppError> {
    let (user_id, root) = user_ctx(&state, &session, &headers).await?;
    let rel = safe_rel_path(&q.path)?;
    let path = root.join(&rel);

    check_quota(&state, &root, &path, content.len() as u64)?;

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }

    fs::write(path, content)?;

    // Writing a path makes it a live file again; a stale tombstone would
    // tell sync clients to delete what the user just (re)created.
    clear_tombstone(&state, &user_id, &rel_key(&rel)).await;
    Ok(StatusCode::OK)
}

#[derive(Deserialize)]
pub struct RelPathBody {
    #[serde(rename = "relPath")]
    rel_path: String,
}

pub async fn create_work_folder(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Json(body): Json<RelPathBody>,
) -> Result<StatusCode, AppError> {
    let root = user_root(&state, &session, &headers).await?;
    let rel = safe_rel_path(&body.rel_path)?;
    fs::create_dir_all(root.join(rel))?;
    Ok(StatusCode::OK)
}

#[derive(Deserialize)]
pub struct DeleteQuery {
    path: String,
    #[serde(rename = "isDir")]
    is_dir: bool,
}

pub async fn delete_work_entry(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Query(q): Query<DeleteQuery>,
) -> Result<StatusCode, AppError> {
    let (user_id, root) = user_ctx(&state, &session, &headers).await?;
    let rel = safe_rel_path(&q.path)?;
    let path = root.join(&rel);
    let key = rel_key(&rel);

    if q.is_dir {
        let mut inside = Vec::new();
        collect_files_under(&path, &key, &mut inside);
        fs::remove_dir_all(path)?;
        for file_rel in inside {
            record_tombstone(&state, &user_id, &file_rel).await;
        }
    } else {
        fs::remove_file(path)?;
        record_tombstone(&state, &user_id, &key).await;
    }
    Ok(StatusCode::OK)
}

#[derive(Deserialize)]
pub struct MoveBody {
    from: String,
    to: String,
}

pub async fn move_work_entry(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
    Json(body): Json<MoveBody>,
) -> Result<StatusCode, AppError> {
    let (user_id, root) = user_ctx(&state, &session, &headers).await?;
    let from_rel = safe_rel_path(&body.from)?;
    let to_rel = safe_rel_path(&body.to)?;
    let from = root.join(&from_rel);
    let to = root.join(&to_rel);
    let from_key = rel_key(&from_rel);
    let to_key = rel_key(&to_rel);

    // A move is a delete at `from` plus a create at `to`, tombstone-wise:
    // stale copies of the old path must not resurface, and any old tombstone
    // on the destination must not kill the file that now lives there.
    let mut moved_files = Vec::new();
    if from.is_dir() {
        collect_files_under(&from, &from_key, &mut moved_files);
    } else {
        moved_files.push(from_key.clone());
    }

    if let Some(parent) = to.parent() {
        fs::create_dir_all(parent)?;
    }

    fs::rename(from, to)?;

    for file_rel in moved_files {
        record_tombstone(&state, &user_id, &file_rel).await;
        let dest = format!("{}{}", to_key, &file_rel[from_key.len()..]);
        clear_tombstone(&state, &user_id, &dest).await;
    }
    Ok(StatusCode::OK)
}

#[derive(Serialize)]
pub struct DeletedEntry {
    #[serde(rename = "relPath")]
    rel_path: String,
    #[serde(rename = "deletedAt")]
    deleted_at: i64,
}

/// Tombstone listing for sync clients: files this workspace deleted, with
/// when. Lets a client holding an unknown local copy distinguish "deleted
/// elsewhere, drop it" from "created locally, push it".
pub async fn list_deleted(
    State(state): State<AppState>,
    session: Session,
    headers: HeaderMap,
) -> Result<Json<Vec<DeletedEntry>>, AppError> {
    let (user_id, _) = user_ctx(&state, &session, &headers).await?;
    let rows = sqlx::query_as::<_, (String, i64)>(
        "SELECT rel_path, deleted_at FROM deleted_files WHERE user_id = ?",
    )
    .bind(&user_id)
    .fetch_all(&state.db)
    .await?;
    Ok(Json(
        rows.into_iter()
            .map(|(rel_path, deleted_at)| DeletedEntry { rel_path, deleted_at })
            .collect(),
    ))
}

File: src/workspace.cpp
#include "workspace.h"

#include <drogon/HttpAppFramework.h>

#include <algorithm>
#include <chrono>
#include <vector>

#include "auth.h"
#include "error.h"
#include "http.h"
#include "state.h"
#include "util.h"

namespace fs = std::filesystem;
using json = nlohmann::json;

namespace lectern::workspace {

namespace {

/// Largest single work file accepted. The Rust build originally inherited
/// axum's 2MB default here, which a .mdn notebook — freehand ink, imported
/// PDF pages, images — passes routinely, so the first sync of such a
/// workspace failed with 413. WORKSPACE_QUOTA_BYTES still caps the account.
size_t max_work_file_bytes()
{
    static const size_t value = static_cast<size_t>(
        util::env_i64("MAX_WORK_FILE_BYTES", 32LL * 1024 * 1024));
    return value;
}

uint64_t modified_ms(const fs::path &path)
{
    std::error_code ec;
    const auto time = fs::last_write_time(path, ec);
    if (ec)
    {
        return 0;
    }
    // file_time_type's epoch is unspecified. std::chrono::clock_cast is the
    // clean way across, but libstdc++ does not implement it for file_clock
    // -- GCC 12 on bookworm, which is what the container builds with -- so
    // shift by the offset between the two clocks instead. The two now()
    // calls land microseconds apart and this returns whole milliseconds.
    const auto system_time =
        std::chrono::time_point_cast<std::chrono::system_clock::duration>(
            time - fs::file_time_type::clock::now() +
            std::chrono::system_clock::now());
    const auto since_epoch = system_time.time_since_epoch();
    const auto millis =
        std::chrono::duration_cast<std::chrono::milliseconds>(since_epoch)
            .count();
    return millis < 0 ? 0 : static_cast<uint64_t>(millis);
}

bool contains(const std::array<std::string_view, 8> &haystack,
              std::string_view needle)
{
    return std::find(haystack.begin(), haystack.end(), needle) !=
           haystack.end();
}

/// Total bytes stored under a workspace root. Walks the tree rather than
/// tracking a running total in the DB: workspaces are small (text files) and
/// this way the number can't drift out of sync with what's on disk.
uint64_t dir_size_bytes(const fs::path &dir)
{
    std::error_code ec;
    uint64_t total = 0;
    for (fs::recursive_directory_iterator it(dir, ec), end; it != end;
         it.increment(ec))
    {
        if (ec)
        {
            break;
        }
        std::error_code stat_ec;
        if (it->is_regular_file(stat_ec) && !stat_ec)
        {
            total += fs::file_size(it->path(), stat_ec);
        }
    }
    return total;
}

// ── Deletion tombstones (see migrations/0007_deleted_files.sql) ──

void record_tombstone(const std::string &user_id, const std::string &rel_path)
{
    try
    {
        state().db->execute(
            "INSERT INTO deleted_files (user_id, rel_path, deleted_at)"
            " VALUES (?, ?, ?)"
            " ON CONFLICT(user_id, rel_path)"
            " DO UPDATE SET deleted_at = excluded.deleted_at",
            {DbValue(user_id), DbValue(rel_path), DbValue(util::now_ms())});
    }
    catch (const std::exception &)
    {
        // Best-effort, exactly as the Rust original: a failed tombstone must
        // not turn a successful delete into an error response.
    }
}

void clear_tombstone(const std::string &user_id, const std::string &rel_path)
{
    try
    {
        state().db->execute(
            "DELETE FROM deleted_files WHERE user_id = ? AND rel_path = ?",
            {DbValue(user_id), DbValue(rel_path)});
    }
    catch (const std::exception &)
    {
    }
}

/// Every file (not directory) under `dir`, as rel keys prefixed with
/// `rel_prefix` — used to tombstone the contents of a deleted/moved folder.
void collect_files_under(const fs::path &dir,
                         const std::string &rel_prefix,
                         std::vector<std::string> &out)
{
    std::error_code ec;
    for (const auto &entry : fs::directory_iterator(dir, ec))
    {
        if (ec)
        {
            return;
        }
        const std::string name = entry.path().filename().string();
        const std::string rel =
            rel_prefix.empty() ? name : rel_prefix + "/" + name;

        std::error_code kind_ec;
        if (entry.is_directory(kind_ec) && !kind_ec)
        {
            collect_files_under(entry.path(), rel, out);
        }
        else
        {
            out.push_back(rel);
        }
    }
}

}  // namespace

std::string safe_rel_path(std::string_view rel_path)
{
    const auto invalid = [] {
        return AppError::bad_request("invalid path");
    };

    if (rel_path.empty())
    {
        throw invalid();
    }
    // A leading separator is an absolute path; Rust's Component::RootDir.
    if (rel_path.front() == '/' || rel_path.front() == '\\')
    {
        throw invalid();
    }

    std::vector<std::string> segments;
    std::string current;
    const auto flush = [&] {
        if (current.empty())
        {
            return;  // "a//b" collapses, same as Path::components()
        }
        if (current == ".")
        {
            current.clear();
            return;  // Component::CurDir is skipped
        }
        if (current == "..")
        {
            throw invalid();  // Component::ParentDir is refused outright
        }
        // A colon is either a Windows drive prefix ("C:") or an NTFS
        // alternate data stream ("file.mdp:evil"); neither belongs in a
        // workspace-relative path.
        if (current.find(':') != std::string::npos)
        {
            throw invalid();
        }
        segments.push_back(current);
        current.clear();
    };

    for (const char c : rel_path)
    {
        if (c == '/' || c == '\\')
        {
            flush();
        }
        else if (c == '\0')
        {
            throw invalid();
        }
        else
        {
            current.push_back(c);
        }
    }
    flush();

    if (segments.empty())
    {
        throw invalid();
    }

    std::string out = segments.front();
    for (size_t i = 1; i < segments.size(); ++i)
    {
        out += "/";
        out += segments[i];
    }
    return out;
}

bool is_work_file(std::string_view rel_path)
{
    return contains(kWorkFileExtensions, util::extension_of(rel_path));
}

json walk_work_dir(const fs::path &dir, const std::string &rel_prefix)
{
    std::error_code ec;
    fs::directory_iterator it(dir, ec);
    if (ec)
    {
        throw AppError::not_found();
    }

    struct Entry
    {
        std::string name;
        bool is_dir = false;
        json value;
    };
    std::vector<Entry> entries;

    for (const auto &entry : it)
    {
        const std::string name = entry.path().filename().string();
        const std::string rel_path =
            rel_prefix.empty() ? name : rel_prefix + "/" + name;

        std::error_code kind_ec;
        const bool is_dir = entry.is_directory(kind_ec);
        if (kind_ec)
        {
            continue;
        }

        if (is_dir)
        {
            entries.push_back({name,
                               true,
                               json{{"name", name},
                                    {"relPath", rel_path},
                                    {"isDir", true},
                                    {"modified", 0},
                                    {"hash", ""},
                                    {"children",
                                     walk_work_dir(entry.path(), rel_path)}}});
            continue;
        }

        // Every regular file is listed, not only the eight app formats. This
        // listing is also what a sync client reads to learn remote state, so
        // anything hidden here but present on disk looks deleted on the next
        // pass -- and the client would dutifully remove its local copy.
        // Dotfiles stay out: they are sync metadata, never user content.
        if (!name.empty() && name.front() == '.')
        {
            continue;
        }

        entries.push_back(
            {name,
             false,
             json{{"name", name},
                  {"relPath", rel_path},
                  {"isDir", false},
                  {"modified", modified_ms(entry.path())},
                  // sha256 hex digest of the file's bytes, empty for
                  // directories. Lets sync clients tell "did this change on
                  // the server" from one listing call without downloading
                  // every file's content.
                  {"hash", util::sha256_file(entry.path())},
                  {"children", json::array()}}});
    }

    std::sort(entries.begin(),
              entries.end(),
              [](const Entry &a, const Entry &b) {
                  if (a.is_dir != b.is_dir)
                  {
                      return a.is_dir;  // directories first
                  }
                  return util::to_lower(a.name) < util::to_lower(b.name);
              });

    json out = json::array();
    for (auto &entry : entries)
    {
        out.push_back(std::move(entry.value));
    }
    return out;
}

json stat_file(const fs::path &path)
{
    std::error_code ec;
    if (!fs::is_regular_file(path, ec) || ec)
    {
        throw AppError::not_found();
    }
    return json{{"modified", modified_ms(path)},
                {"hash", util::sha256_file(path)}};
}

void check_quota(const fs::path &root,
                 const fs::path &target,
                 uint64_t incoming_len)
{
    const uint64_t quota = state().workspace_quota_bytes;
    if (quota == 0)
    {
        return;
    }

    std::error_code ec;
    const uint64_t existing_len =
        fs::is_regular_file(target, ec) ? fs::file_size(target, ec) : 0;
    if (ec || incoming_len <= existing_len)
    {
        return;
    }

    const uint64_t projected =
        dir_size_bytes(root) - existing_len + incoming_len;
    if (projected > quota)
    {
        throw AppError::bad_request(
            "workspace is full (" + std::to_string(quota / (1024 * 1024)) +
            " MB max) — delete something first");
    }
}

fs::path user_root(const drogon::HttpRequestPtr &req, std::string *user_id_out)
{
    const std::string user_id = auth::current_user_id_with_headers(req);
    if (user_id_out != nullptr)
    {
        *user_id_out = user_id;
    }
    const fs::path root = state().workspaces_dir / user_id;
    std::error_code ec;
    fs::create_directories(root, ec);
    return root;
}

void register_routes()
{
    auto &app = drogon::app();

    app.registerHandler(
        "/api/workspace",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                return http::json_response(
                    walk_work_dir(user_root(req), ""));
            });
        },
        {drogon::Get});

    app.registerHandler(
        "/api/workspace/stat",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                const fs::path root = user_root(req);
                const std::string rel =
                    safe_rel_path(http::required_query(req, "path"));
                return http::json_response(stat_file(root / rel));
            });
        },
        {drogon::Get});

    app.registerHandler(
        "/api/workspace/file",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&]() -> drogon::HttpResponsePtr {
                std::string user_id;
                const fs::path root = user_root(req, &user_id);
                const std::string rel =
                    safe_rel_path(http::required_query(req, "path"));
                const fs::path path = root / rel;

                if (req->method() == drogon::Get)
                {
                    const auto content = util::read_file(path);
                    if (!content)
                    {
                        throw AppError::not_found();
                    }
                    return http::text_response(*content);
                }

                // A view, not a copy: Drogon already holds the whole
                // body in memory, and copying it doubled the peak cost
                // of every upload -- which matters once
                // MAX_WORK_FILE_BYTES is raised for large attachments.
                const std::string_view content = req->getBody();
                // Drogon's body limit is global and has to accommodate PDF
                // export, so the per-file ceiling is enforced here.
                if (content.size() > max_work_file_bytes())
                {
                    throw AppError::bad_request(
                        "file is too large (" +
                        std::to_string(max_work_file_bytes() / (1024 * 1024)) +
                        " MB max)");
                }
                check_quota(root, path, content.size());
                util::write_file_atomic(path, content);

                // Writing a path makes it a live file again; a stale
                // tombstone would tell sync clients to delete what the user
                // just (re)created.
                clear_tombstone(user_id, rel);
                return http::ok_response();
            });
        },
        {drogon::Get, drogon::Put});

    app.registerHandler(
        "/api/workspace/folder",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                const fs::path root = user_root(req);
                const std::string rel = safe_rel_path(
                    http::json_string(http::json_body(req), "relPath"));
                std::error_code ec;
                fs::create_directories(root / rel, ec);
                if (ec)
                {
                    throw AppError::internal(ec.message());
                }
                return http::ok_response();
            });
        },
        {drogon::Post});

    app.registerHandler(
        "/api/workspace/entry",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                std::string user_id;
                const fs::path root = user_root(req, &user_id);
                const std::string rel =
                    safe_rel_path(http::required_query(req, "path"));
                const fs::path path = root / rel;
                const bool is_dir = http::query_bool(req, "isDir");

                std::error_code ec;
                if (is_dir)
                {
                    std::vector<std::string> inside;
                    collect_files_under(path, rel, inside);
                    if (fs::remove_all(path, ec) == 0 || ec)
                    {
                        throw AppError::not_found();
                    }
                    for (const auto &file_rel : inside)
                    {
                        record_tombstone(user_id, file_rel);
                    }
                }
                else
                {
                    if (!fs::remove(path, ec) || ec)
                    {
                        throw AppError::not_found();
                    }
                    record_tombstone(user_id, rel);
                }
                return http::ok_response();
            });
        },
        {drogon::Delete});

    app.registerHandler(
        "/api/workspace/move",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                std::string user_id;
                const fs::path root = user_root(req, &user_id);
                const auto body = http::json_body(req);
                const std::string from_key =
                    safe_rel_path(http::json_string(body, "from"));
                const std::string to_key =
                    safe_rel_path(http::json_string(body, "to"));
                const fs::path from = root / from_key;
                const fs::path to = root / to_key;

                // A move is a delete at `from` plus a create at `to`,
                // tombstone-wise: stale copies of the old path must not
                // resurface, and any old tombstone on the destination must
                // not kill the file that now lives there.
                std::vector<std::string> moved_files;
                std::error_code ec;
                if (fs::is_directory(from, ec) && !ec)
                {
                    collect_files_under(from, from_key, moved_files);
                }
                else
                {
                    moved_files.push_back(from_key);
                }

                const fs::path parent = to.parent_path();
                if (!parent.empty())
                {
                    fs::create_directories(parent, ec);
                }

                fs::rename(from, to, ec);
                if (ec)
                {
                    throw AppError::not_found();
                }

                for (const auto &file_rel : moved_files)
                {
                    record_tombstone(user_id, file_rel);
                    const std::string dest =
                        to_key + file_rel.substr(from_key.size());
                    clear_tombstone(user_id, dest);
                }
                return http::ok_response();
            });
        },
        {drogon::Post});

    // Tombstone listing for sync clients: files this workspace deleted, with
    // when. Lets a client holding an unknown local copy distinguish "deleted
    // elsewhere, drop it" from "created locally, push it".
    app.registerHandler(
        "/api/workspace/deleted",
        [](const drogon::HttpRequestPtr &req, HttpCallback &&callback) {
            http::guard(std::move(callback), [&] {
                std::string user_id;
                user_root(req, &user_id);
                const auto rows = state().db->query_all(
                    "SELECT rel_path, deleted_at FROM deleted_files"
                    " WHERE user_id = ?",
                    {DbValue(user_id)});

                json out = json::array();
                for (const auto &row : rows)
                {
                    out.push_back({{"relPath", row.text(0)},
                                   {"deletedAt", row.i64(1)}});
                }
                return http::json_response(out);
            });
        },
        {drogon::Get});
}

}  // namespace lectern::workspace
